pooch/core.py

```diff
@@ -669,23 +693,42 @@ def download_action(path, known_hash):
     return action, verb
 
 
-def stream_download(url, fname, known_hash, downloader, pooch=None):
+def stream_download(url, fname, known_hash, downloader, pooch=None, retry_if_failed=0):
     """
     Stream the file and check that its hash matches the known one.
 
     The file is first downloaded to a temporary file name in the cache folder.
     It will be moved to the desired file name only if the hash matches the
     known hash. Otherwise, the temporary file is deleted.
 
+    If the download fails for either a bad connection or a hash mismatch, we
+    will retry the download the specified number of times in case the failure
+    was due to a network error.
     """
     # Ensure the parent directory exists in case the file is in a subdirectory.
     # Otherwise, move will cause an error.
     if not fname.parent.exists():
         os.makedirs(str(fname.parent))
 
-    # Stream the file to a temporary so that we can safely check its hash
-    # before overwriting the original.
-    with temporary_file(path=str(fname.parent)) as tmp:
-        downloader(url, tmp, pooch)
-        hash_matches(tmp, known_hash, strict=True, source=str(fname.name))
-        shutil.move(tmp, str(fname))
+    download_attempts = 1 + retry_if_failed
+
+    for i in range(download_attempts):
+        try:
+            # Stream the file to a temporary so that we can safely check its
+            # hash before overwriting the original.
+            with temporary_file(path=str(fname.parent)) as tmp:
+                downloader(url, tmp, pooch)
+                hash_matches(tmp, known_hash, strict=True, source=str(fname.name))
+                shutil.move(tmp, str(fname))
+            break
+        except (ValueError, requests.exceptions.RequestException):
+            if i == download_attempts - 1:
+                raise
+            retries_left = download_attempts - (i + 1)
+            get_logger().info(
+                "Failed to download '%s'. "
+                "Will attempt the download again %d time%s.",
+                str(fname.name),
+                retries_left,
+                "s" if retries_left > 1 else "",
+            )
```

Comment:
Maybe put a a small sleep here?

Network issues sometimes need a little bit of time to resolve.

---

It may not be necessary now, but another idea to consider in the future is to do something like wget, which uses linear backoff between multiple waits.

https://www.gnu.org/software/wget/manual/wget.html

Reply:
That's a good idea. From the wget man page:

> If you don’t want Wget to wait between every retrieval, but only between retries of failed downloads, you can use this option. Wget will use linear backoff, waiting 1 second after the first failure on a given file, then waiting 2 seconds after the second failure on that file, up to the maximum number of seconds you specify.
> By default, Wget will assume a value of 10 seconds. 

Those seem like good defaults (wait 1s more every time up to 10s). What do you think?

Reply:
Yep, sounds good!

Reply:
Alright, I've implemented this in  733c5fc . Seems to work fine and doesn't take too long between retries.